runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -831,6 +831,143 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 		return getSpineSize(J9GC_J9OBJECT_CLAZZ(arrayPtr, this), layout, getSizeInElements(arrayPtr));
 	}
 
+#if defined(J9VM_ENV_DATA64)
+	/**
+	 * Sets data pointer of a contiguous indexable object.
+	 * Sets the data pointer of a contiguous indexable object; in this case
+	 * dataAddr will point directly into the data right after dataAddr field
+	 * (data resides in heap).
+	 *
+	 * @param arrayPtr      Pointer to the indexable object whose size is required
+	 */
+	MMINLINE void
+	setDataAddrForContiguous(J9IndexableObject *arrayPtr)
+	{
+		AssertContiguousArrayletLayout(arrayPtr);
+		bool const compressed = compressObjectReferences();
+		U_64 address = (U_64)((UDATA)arrayPtr + contiguousHeaderSize());
+
+		if (compressed) {
+			((J9IndexableObjectContiguousCompressed *)arrayPtr)->dataAddr = address;
+		} else {
+			((J9IndexableObjectContiguousFull *)arrayPtr)->dataAddr = address;
+		}
```

Comment:
You could share the code for setter and getter.... have dataAddrSlotForContiguous() (and one for Discontiguous) that will give you address of the dataAddr Slot, and then setter and getter just de-references the address.